### PR TITLE
Fix some Debuginfo.t scopes in the frontend

### PR DESCRIPTION
**Pull request by mshinwell**
This fixes some places in the frontend where the scoping information in `Debuginfo.t` was not accurately identifying what was going on:
- during eta-expansion of primitives
- during expansion of `lazy`
- during `Translcore.transl_apply.build_apply` which I think is related to optional arguments (I'm unsure about this one).

There is also at least one case elsewhere relating to partial application wrapper creation where the scoping information is wrong.  That will be dealt with in a later patch due to a dependency.

Originally by @basimkhajwal

@stedolan could you please look at this?

### CODE REVIEW

**Comment by mshinwell** [comment]
This causes diffs in backtraces (e.g. in the `lazy` tests) of the form:
```
2021-09-13T10:21:46.9223462Z -Raised at Lazy.l1 in file "lazy.ml", line 7, characters 28-45
2021-09-13T10:21:46.9224360Z +Raised at Lazy.l1.(fun) in file "lazy.ml", line 7, characters 28-45
```
I wonder whether we should have a distinguished scope kind for lambdas introduced by `lazy`.

**Comment by stedolan** [comment]
The changes look good to me, but there should also be some tests of the new functionality in `ocaml/testsuite/tests/backtrace/names.ml`.

The change in build_apply is to do with labeled arguments that have been permuted and abstracted over:
```
let f ~x ~y = Printf.printf "%d %d\n" x y
let g = f ~y:5
```
Here `g` defines a closure `fun ~x -> f ~x ~y:5`. I don't have a strong opinion about this one: calling this closure an anonymous function is fine, but not giving it a name also seems fine (this is what happens for the thunks introduced by caml_apply currently).

(It's possible that these functions can never appear in backtraces, if they immediately tail-call the underlying `f`. If that's the case, it's fine not to include a test for them, but the other cases should have one)

I agree that Lazy should have a new scope type. It's confusing to see `M.foo.(fun)` when `M.foo` contains no `fun`.

**Comment by mshinwell** [comment]
I think it's better if the scopes match up with the structure of the lambdas, but they don't necessarily have to be displayed to the user.  How about adding a "partial application wrapper" case?  That would deal with the labelled arguments case and also the cases I mentioned above for a subsequent patch.  It could be elided upon printing.

**Comment by stedolan** [comment]
A partial application wrapper case `Foo.bar.(partial)` makes sense to me! (I think it is appropriate to use this for the eta-expansion wrappers of primitives too, even though technically eta-expansion is something else. Distinguishing the two would be more confusing).

I don't think it's necessary to elide these on printing. That elision sounds like it might be more confusing than having them appear. I don't think they'll show up very much in real backtraces, as they usually tail-call the underlying function.

File: ocaml/testsuite/tests/backtrace/names.ml [hunk 1]
```diff
@@ -97,6 +102,16 @@ let inline_object f =
   end in
   obj#meth
 
+let[@inline never] labelled_arguments_partial k =
+  let f ~a = ignore a; k (); fun ~b -> ignore b; () in
```
Comment (stedolan) [inline ocaml/testsuite/tests/backtrace/names.ml:106]: It looks like this `f` is getting inlined, ruining the effect. Try adding a noinline or opaque_identity on `f`.

File: ocaml/lambda/debuginfo.ml [hunk 2]
```diff
@@ -72,6 +78,14 @@ module Scoped_location = struct
     in
     cons Sc_method_definition str
 
+(*
+  let enter_lazy ~scopes = cons Sc_lazy (str scopes)
+*)
+  let enter_lazy ~scopes = cons Sc_lazy (dot scopes "(lazy)")
```
Comment (stedolan) [inline ocaml/lambda/debuginfo.ml:84]: This is adding a double-paren `((lazy))`. (cf what enter_anonymous_function does)

Comment (mshinwell) [reply ocaml/lambda/debuginfo.ml:84]: Ah, the commented code will be the final version -- this is just for debugging to make sure the scope is there.

**Comment by mshinwell** [comment]
What I don't understand yet though is why `(partial)` doesn't appear in the reference file.

File: ocaml/lambda/debuginfo.ml (hunk 2 above)
Comment (stedolan) [reply ocaml/lambda/debuginfo.ml:84]: Surely the scope should be there regardless?

**Comment by stedolan** [comment]
> What I don't understand yet though is why (partial) doesn't appear in the reference file.

See the comment about inlining. (I think inlining is optimising away the closure)

**Comment by mshinwell** [comment]
@stedolan There are three other cases though: the exception raise, `apply` and `revapply`?

File: ocaml/lambda/debuginfo.ml (hunk 2 above)
Comment (mshinwell) [reply ocaml/lambda/debuginfo.ml:84]: Sure, but I thought the idea is that these wouldn't be in the textual version of the scope?

File: ocaml/testsuite/tests/backtrace/names.ml (hunk 1 above)
Comment (mshinwell) [reply ocaml/testsuite/tests/backtrace/names.ml]: This doesn't seem to have worked either.  I am still pondering why these partial ones don't seem to work (also see below).

Comment (mshinwell) [reply ocaml/testsuite/tests/backtrace/names.ml]: Ah, the problem is that the location needs to be passed around the loop of `build_apply` I think.  I'm trying to fix this.

**Comment by mshinwell** [comment]
OK this is now successfully exercising the two new kinds of scopes in three different cases (lazy, partial application, eta expansion of primitives).  I couldn't get any `%apply` / `%revapply` eta expansion tests to work, but I don't think they will improve test coverage anyway (eta expansion is already checked with `raise`), and enough time has been spent on this.

**Review by stedolan — APPROVED** [review]
Looks good to me (but there are some CI failures)

**Comment by lpw25** [comment]
Fix here: https://github.com/ocaml-flambda/ocaml-jst/pull/15

**Comment by lpw25** [comment]
(I seem to have lost a comment I wrote this morning, repeated here for completeness)

This PR breaks `__FUNCTION__`, causing it to always have a `(partial)` suffix, as evidenced by the changes in this PR to the `ocaml/testsuite/tests/translprim/locs.reference` test.

**Pull request merged**